compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,220 @@
+    /// Otherwise returns false. The given instruction is not removed from the function.
+    fn optimize_const_brillig_call(
+        &mut self,
+        instruction_id: InstructionId,
+        brillig_functions: &BTreeMap<FunctionId, Function>,
+        brillig_functions_we_could_not_inline: &mut HashSet<FunctionId>,
+        inliner_aggressiveness: i64,
+        error_selector_to_type: &BTreeMap<ErrorSelector, Type>,
+    ) -> bool {
+        let instruction = &self.dfg[instruction_id];
+        let Instruction::Call { func: func_id, arguments } = instruction else {
+            return false;
+        };
+
+        let func_value = &self.dfg[*func_id];
+        let Value::Function(func_id) = func_value else {
+            return false;
+        };
+
+        let Some(function) = brillig_functions.get(func_id) else {
+            return false;
+        };
+
+        if !arguments.iter().all(|argument| self.dfg.is_constant(*argument)) {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        // The function we have is already a copy of the original function, but we need to clone
+        // it again because there might be multiple calls to the same brillig function.
+        let mut function = Function::clone_with_id(*func_id, function);
+
+        // Find the entry block and remove its parameters
+        let entry_block_id = function.entry_block();
+        let entry_block = &mut function.dfg[entry_block_id];
+        let entry_block_parameters = entry_block.take_parameters();
+
+        assert_eq!(arguments.len(), entry_block_parameters.len());
+
+        // Replace the ValueId of parameters with the ValueId of arguments
+        for (parameter_id, argument_id) in entry_block_parameters.iter().zip(arguments) {
+            // Lookup the argument in the current function and insert it in the function copy
+            let new_argument_id = copy_constant_to_function(self, &mut function, *argument_id);
+            function.dfg.set_value_from_id(*parameter_id, new_argument_id);
+        }
+
+        // Try to fully optimize the function. If we can't, we can't inline it's constant value.
+        let Ok(mut function) = optimize(function, inliner_aggressiveness, error_selector_to_type)
+        else {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        };
+
+        let entry_block = &mut function.dfg[entry_block_id];
+
+        // If the entry block has instructions, we can't inline it (we need a terminator)
+        if !entry_block.instructions().is_empty() {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        let terminator = entry_block.take_terminator();
+        let TerminatorInstruction::Return { return_values, call_stack: _ } = terminator else {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        };
```

Comment:
Just to confirm: the presence of these could not be verified before trying to inline everything, because the act of inlining could have changed them?

Reply:
Correct. Or, well, in addition to inlining, jumps and stuff like that is all inlined into a single block.